### PR TITLE
Separate translations from XML, PO files

### PR DESCRIPTION
**Pull request by sbcgua**
See the concept [here](https://github.com/abapGit/abapGit/issues/2539#issuecomment-1495537247)
And also https://www.gnu.org/software/gettext/manual/gettext.html

Creating it as wip for the moment, but this is a fully functional code, can/should be tested and discussed. Minor tweaks (comment removal) to be done before merging (to fix the linter messages in particular).

See the sample repo here: https://github.com/sbcgua/i18n_test

See also the "Challenges / todo / doubts" section in the Concept above. These are the next steps for next PRs. But this one is probably complete.

## Technical notes

- `zcl_abapgit_objects_files` accepts/gives out i18n files
  - files are now also expose `i18n_params`. Ugly. Needs to be improved in future. Triplicates xml_in/out code. Fortunatelly, the code is identical so it is more a question of where should it belong
- `zcl_abapgit_po_file` and `zif_abapgit_i18n_file` is the heart - IO for the PO files
- `zcl_abapgit_lxe_texts` mainly rewiring related to the new class
- rewiring in `objects_super` - to be discussed, also looks like some major refactoring needed to access files in a leaner way (and not just call `lxe_to_po` with the files param - which is also an option).

### CODE REVIEW

**Review by larshp — CHANGES_REQUESTED** [review]
need this coordinated with https://github.com/SAP/abap-file-formats/pull/511

**Comment by larshp** [comment]
suggest @schneidermic0 to schedule a call

btw, found https://marketplace.visualstudio.com/items?itemName=mrorz.language-gettext

(I'd personally like to have all structured data in JSON, instead of "new" file formats, but guess its not gonna happen)

**Comment by schneidermic0** [comment]
Thanks, @larshp for the heads-up: I have just scheduled a call

**Comment by sbcgua** [comment]
I think it ready for review. And testing (I'm going to test it further yet, though the recent changes are mainly refactoring ... lot's of it).

There are many changes but they are for good.

1) A new abstraction - `zcl_abapgit_i18n_params`, which wraps the `zif_abapgit_definitions=>ty_i18n_params` + some extra conveniences and place for relevant utilities
2) serialization and deserialization of LXE is moved to `zcl_abapgit_objects`, so called in one place and declutter the object classes (:tada:)
3) the deserialization is suboptimal ... I'm probably going to do some refactoring over the serialization and deserialization - it is very difficult to change and is scattered across many places. Room for improvement. Then I'll adjust the LXE deserialization to the right place (to run after the activation)
4) old XML-based format. I kept the code for now. Though it is not executed. In my opinion, it must be deleted - no one uses it to my knowledge. And it actually has a subtle bug (below) if run between the system releases.
5) There is one known bug in the PO approach. Different releases have different LXE keys (!). In fact, due to the way PO is implemented, it does not impact deserialization (unlike the XML approach). But it produces unnecessary diffs. Fortunately, LXE has still the experimental status which gives some freedom to collect more real statistics on that.

P.S. (to 2) I also think that it makes sense to expose `serialize_texts` into the interface, to make the call in one place (objects_serialize). This would make the code cleaner. Now the decision about the text serialization is not centralized. Maybe also `serialize_longtexts` - this will also make it easier to move longtexts to PO later.

**Comment by sbcgua** [comment]
not sure why tests are failing, looks like it doesn't find a definition of an i18n interface

**Comment by mbtools** [comment]
add `zif_abapgit_i18n...` to `input_filter` in https://github.com/abapGit/abapGit/blob/main/test/abap_transpile.json

**Comment by sbcgua** [comment]
> add `zif_abapgit_i18n...` to `input_filter` in https://github.com/abapGit/abapGit/blob/main/test/abap_transpile.json

didn't work :/

```
RUNNING

node:internal/process/esm_loader:46
      internalBinding('errors').triggerUncaughtException(
                                ^
Void type: LXECUSTMNR
(Use `node --trace-uncaught ...` to show where the exception was thrown)
```

**Comment by mbtools** [comment]
there are a few DDIC objects that the transpiler doesn't know: `lxecustmnr`, `lxeisolang`, `lxeobjname`, `lxe_pcx_s1`. either these need to be added to the `deps` folder or try to exclude the unit tests that reference them in `abap_transpile.json`.

**Comment by sbcgua** [comment]
UT seems to work now. Lars, Marc, thanks for hints

**Comment by larshp** [comment]
I'm okay with the changes, ok to merge from my POV

File: src/objects/zcl_abapgit_objects.clas.abap [hunk 1]
```diff
@@ -595,9 +601,7 @@ CLASS zcl_abapgit_objects IMPLEMENTATION.
       RETURN.
     ENDIF.
 
-    li_obj = create_object( is_item     = is_item
-                            iv_language = zif_abapgit_definitions=>c_english ).
-
+    li_obj = create_object( is_item ).
```
Comment (mbtools) [inline src/objects/zcl_abapgit_objects.clas.abap:604]: thanks for simplifying `create_object`. i had that on my todo as well.

**Review by mbtools — COMMENTED** [review]
Just a question if we can add other code and ddic object types (see above)

File: src/objects/zcl_abapgit_objects.clas.abap (hunk 1 above)
Comment (sbcgua) [reply src/objects/zcl_abapgit_objects.clas.abap:604]: I'm planning some refactorings in that area too ... maybe we should synchronize the plans a bit ...

File: src/objects/texts/zcl_abapgit_lxe_texts.clas.abap [hunk 2]
```diff
@@ -197,6 +198,23 @@ CLASS ZCL_ABAPGIT_LXE_TEXTS IMPLEMENTATION.
   ENDMETHOD.
 
 
+  METHOD class_constructor.
+
+    APPEND 'CLAS' TO gt_supported_obj_types.
```
[the remark this answers is not in the document]
Comment (sbcgua) [reply src/objects/texts/zcl_abapgit_lxe_texts.clas.abap:203]: INTFs do not have texts ... or ? At least they don't in my products so I skipped them. Essentially I added the objects where I have an appropriate sample and usage. It is the only reason for the list as it is. So if you know some object has texts and can check it, then of course feel free to extend the list. It's specifically created this way to be easily extended. :)

[the remark this answers is not in the document]
Comment (mbtools) [reply src/objects/texts/zcl_abapgit_lxe_texts.clas.abap:203]: They all have descriptions that can be translated (and intf has methods, parameters, ...). LXE is generic. That's the beauty of it. `get_lxe_object_list` should tell you if there are texts for an object. Why do we need the list of supported objects? Can't we do all with this?

[the remark this answers is not in the document]
Comment (sbcgua) [reply src/objects/texts/zcl_abapgit_lxe_texts.clas.abap:203]: > LXE is generic

Yes, but I'd prefer doing it in a controllable fashion. Step by step. This step is already quite large and it is focused on the technology itself. But I really have no objections to an extension after this PR

**Comment by sbcgua** [comment]
If no objections I'd merge this in the nearest days.

**Pull request merged**